C/plugins/common/http_sender.cpp

```diff
@@ -35,64 +35,71 @@ HttpSender::~HttpSender()
 }
 
 /**
- * @brief Creates the "debug-trace" directory under the base directory returned by getDataDir().
+ * @brief Creates the '/logs/debug-trace' directory under the directory returned by getDataDir().
  * 
- * This function ensures that the "debug-trace" directory is created only if the base directory exists
+ * This function ensures that both the 'logs' directory and the 'debug-trace' directory are created if they do not exist.
  */
 bool HttpSender::createDebugTraceDirectory() 
 {
-    // Step 1: Retrieve the base data directory
-    std::string baseDir = getDataDir(); 
-    std::string debugTraceDir = baseDir + "/debug-trace";
+    // Retrieve the 'logs' and 'debug-trace' directory paths
+    std::string logsDir = getDataDir() + "/logs";
+    std::string debugTraceDir = logsDir + "/debug-trace";
 
-    // Step 2: Check if the base directory exists
-    struct stat baseInfo;
-    if (stat(baseDir.c_str(), &baseInfo) == 0 && (baseInfo.st_mode & S_IFDIR))
-    {
-        // Base directory exists, proceed to creating the debug-trace directory
+    // Ensure path consistency with getDebugTracePath(). Assert is commented out to prevent unexpected runtime interruptions in execution
+    // std::assert(debugTraceDir == getDebugTracePath()); 
 
-        // Step 3: Check if the debug-trace directory exists
-        struct stat debugTraceInfo;
-        if (stat(debugTraceDir.c_str(), &debugTraceInfo) != 0) 
+    auto createDir = [](const std::string& dirPath) -> bool 
```

Comment:
This code is fine but I wanted to point out that there is a method in our common library called _createDirectory_ (see [file_utils.cpp](https://github.com/fledge-iot/fledge/blob/develop/C/common/file_utils.cpp)). I did not comment before because you were doing additional work like setting access permissions. If any of these ideas could be put into _createDirectory_, that would be good to know.

Reply:
There are two issue why I can not use createDirectory (see [file_utils.cpp](https://github.com/fledge-iot/fledge/blob/develop/C/common/file_utils.cpp)).
1. The function do not return status, it throws exception.
2. It do not check if the path is a file or a directory. It do not check this 
            if (dirInfo.st_mode & S_IFDIR)
            {
                return true; // Directory exists
            }

Reply:
As I wrote, you don't need to use it. A couple of points though:

- I don't see why a method throwing an exception disqualifies it. The design was influenced by a scenario in S2OPCUA South in which I needed to create a complex directory subtree, one subdirectory at a time. All paths were created in code so the path strings were always correct and the directory creation was unlikely to fail. This way I could put all _createDirectory()_ calls into a single try block. If anything failed, the failed path would be in the exception string. This way, I don't have check a status code, log an error message and return false for every _createDirectory()_ call. If you only need to create one directory, put _createDirectory()_ in the `try` block and the error log in the `catch`.
- _createDirectory()_ does indeed not check that the path exists but is not a directory. Good idea; this can be added.

I see you went with a _mode_t_ value of 0755 which gives the Other group read and search access but not write. Method _createDirectory()_ uses _S_IRWXU | S_IRWXG_ which translates to 0770 meaning Other gets nothing. Your _mode_t_ setting is probably better. By the way, I found the _mode_t_ symbolic names on the [Gnu website](https://www.gnu.org/software/libc/manual/html_node/Permission-Bits.html). I also found a cool utility while researching this: the [Chmod Calculator](https://chmodcommand.com/)!

Reply:
Thanks, @RVerhoeff , for sharing! It is very insightful.

In the meantime, I was wondering if we could use C++17 in our project. If we did, we would have access to the standard [create_directory](https://en.cppreference.com/w/cpp/filesystem/create_directory).

I am not aware of the background for sticking to C++11.

Reply:
I don't know why either. Upgrading to C++ 17 might be a big project and would have to include the entire Fledge and FogLamp code bases. We will have to ask @MarkRiddoch.